Tests/Tests/STPBankAccountTest.m

```diff
@@ -100,53 +117,82 @@ - (void)testBankAccountEquals {
-    STPBankAccount *bankAccountWithAttributes = [STPBankAccount decodedObjectFromAPIResponse:apiResponse];
-
-    XCTAssertEqualObjects([bankAccountWithAttributes bankAccountId], @"something", @"bankAccountId is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes last4], @"6789", @"last4 is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes bankName], @"STRIPE TEST BANK", @"bankName is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes country], @"US", @"country is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes fingerprint], @"something", @"fingerprint is set correctly");
-    XCTAssertEqualObjects([bankAccountWithAttributes currency], @"usd", @"currency is set correctly");
-    XCTAssertEqual([bankAccountWithAttributes status], STPBankAccountStatusNew);
-    XCTAssertEqualObjects([bankAccountWithAttributes accountHolderName], @"John Doe");
-    XCTAssertEqual([bankAccountWithAttributes accountHolderType], STPBankAccountHolderTypeCompany);
-
-    NSDictionary *allResponseFields = bankAccountWithAttributes.allResponseFields;
-    XCTAssertEqual(allResponseFields[@"foo"], @"bar");
-    XCTAssertEqual(allResponseFields[@"last4"], @"6789");
-    XCTAssertNil(allResponseFields[@"baz"]);
+- (void)testDecodedObjectFromAPIResponseRequiredFields {
+    NSArray<NSString *> *requiredFields = @[
+                                            @"id",
+                                            @"last4",
+                                            @"bank_name",
+                                            @"country",
+                                            @"currency",
+                                            @"status",
+                                            ];
+
+    for (NSString *field in requiredFields) {
+        NSMutableDictionary *response = [[self completeAttributeDictionary] mutableCopy];
+        [response removeObjectForKey:field];
+
+        XCTAssertNil([STPBankAccount decodedObjectFromAPIResponse:response]);
+    }
+
+    XCTAssert([STPBankAccount decodedObjectFromAPIResponse:[self completeAttributeDictionary]]);
+}
+
+- (void)testDecodedObjectFromAPIResponseMapping {
+    NSDictionary *response = [self completeAttributeDictionary];
+    STPBankAccount *bankAccount = [STPBankAccount decodedObjectFromAPIResponse:response];
+
+    XCTAssertEqualObjects(bankAccount.bankAccountId, @"ba_1AXvnKEOD54MuFwSotKc6xq0");
+    XCTAssertEqualObjects(bankAccount.accountHolderName, @"Jane Austen");
+    XCTAssertEqual(bankAccount.accountHolderType, STPBankAccountHolderTypeIndividual);
+    XCTAssertEqualObjects(bankAccount.bankName, @"STRIPE TEST BANK");
+    XCTAssertEqualObjects(bankAccount.country, @"US");
+    XCTAssertEqualObjects(bankAccount.currency, @"usd");
+    XCTAssertEqualObjects(bankAccount.fingerprint, @"C5fW7AwE3of8bHvV");
+    XCTAssertEqualObjects(bankAccount.last4, @"6789");
+    XCTAssertNil(bankAccount.routingNumber);
```

Comment:
Yeah it seems like an oversight that this isn't parsed currently? Not sure if there was some intentional reason we don't parse this (although I assume not because we have a property for it). Also `routingNumber` is declared as nonnull, which is just entirely wrong. Should change that annotation and/or parse it and add it to required keys I guess.

Somewhat relatedly: It's suuuuper weird to me that STPBankAccount and STPCard are subclasses of their own Params objects. Glad we didn't do that for Sources.

What do you think @bg-stripe?